### DISCUSSION TITLE
Daily Temporal Climatology using xcdat

### DISCUSSION THREAD

**Discussion by gkb999** [xCDAT/xcdat#469] [4. General]
Hi there, 
I have daily data for about 40 years, each year have been imported using mf_datset and nested along 'time' dimension.
Trying to use Daily temporal climatology on this, and it shows memory issues.
The error reads as follows:
```python
File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xcdat\temporal.py:340, in TemporalAccessor.group_average(self, data_var, freq, weighted, keep_weights, season_config)
    215 """Returns a Dataset with average of a data variable by time group.
    216 
    217 Parameters
   (...)
    336 }
    337 """
    338 self._set_data_var_attrs(data_var)
--> 340 return self._averager(
    341     data_var,
    342     "group_average",
    343     freq,
    344     weighted=weighted,
    345     keep_weights=keep_weights,
    346     season_config=season_config,
    347 )

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xcdat\temporal.py:777, in TemporalAccessor._averager(self, data_var, mode, freq, weighted, keep_weights, reference_period, season_config)
    775     dv = self._average(dv, time_bounds)
    776 elif self._mode in ["group_average", "climatology", "departures"]:
--> 777     dv = self._group_average(dv, time_bounds)
    779 # The original time dimension is dropped from the dataset because
    780 # it becomes obsolete after the data variable is averaged. When the
    781 # averaged data variable is added to the dataset, the new time dimension
    782 # and its associated coordinates are also added.
    783 ds = ds.drop_dims(self.dim)  # type: ignore

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xcdat\temporal.py:1150, in TemporalAccessor._group_average(self, data_var, time_bounds)
   1145 # Ensure missing data (`np.nan`) receives no weight (zero). To
   1146 # achieve this, first broadcast the one-dimensional (temporal
   1147 # dimension) shape of the `weights` DataArray to the
   1148 # multi-dimensional shape of its corresponding data variable.
   1149 weights, _ = xr.broadcast(self._weights, dv)
-> 1150 weights = xr.where(np.isnan(dv), 0.0, weights)
   1152 # Perform weighted average using the formula
   1153 # WA = sum(data*weights) / sum(weights). The denominator must be
   1154 # included to take into account zero weight for missing data.
   1155 with xr.set_options(keep_attrs=True):

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xarray\core\computation.py:1861, in where(cond, x, y, keep_attrs)
   1858     keep_attrs = _get_keep_attrs(default=False)
   1860 # alignment for three arguments is complicated, so don't support it yet
-> 1861 result = apply_ufunc(
   1862     duck_array_ops.where,
   1863     cond,
   1864     x,
   1865     y,
   1866     join="exact",
   1867     dataset_join="exact",
   1868     dask="allowed",
   1869     keep_attrs=keep_attrs,
   1870 )
   1872 # keep the attributes of x, the second parameter, by default to
   1873 # be consistent with the `where` method of `DataArray` and `Dataset`
   1874 # rebuild the attrs from x at each level of the output, which could be
   1875 # Dataset, DataArray, or Variable, and also handle coords
   1876 if keep_attrs is True and hasattr(result, "attrs"):

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xarray\core\computation.py:1196, in apply_ufunc(func, input_core_dims, output_core_dims, exclude_dims, vectorize, join, dataset_join, dataset_fill_value, keep_attrs, kwargs, dask, output_dtypes, output_sizes, meta, dask_gufunc_kwargs, *args)
   1194 # feed DataArray apply_variable_ufunc through apply_dataarray_vfunc
   1195 elif any(isinstance(a, DataArray) for a in args):
-> 1196     return apply_dataarray_vfunc(
   1197         variables_vfunc,
   1198         *args,
   1199         signature=signature,
   1200         join=join,
   1201         exclude_dims=exclude_dims,
   1202         keep_attrs=keep_attrs,
   1203     )
   1204 # feed Variables directly through apply_variable_ufunc
   1205 elif any(isinstance(a, Variable) for a in args):

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xarray\core\computation.py:303, in apply_dataarray_vfunc(func, signature, join, exclude_dims, keep_attrs, *args)
    298 result_coords, result_indexes = build_output_coords_and_indexes(
    299     args, signature, exclude_dims, combine_attrs=keep_attrs
    300 )
    302 data_vars = [getattr(a, "variable", a) for a in args]
--> 303 result_var = func(*data_vars)
    305 out: tuple[DataArray, ...] | DataArray
    306 if signature.num_outputs > 1:

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xarray\core\computation.py:760, in apply_variable_ufunc(func, signature, exclude_dims, dask, output_dtypes, vectorize, keep_attrs, dask_gufunc_kwargs, *args)
    755     if vectorize:
    756         func = _vectorize(
    757             func, signature, output_dtypes=output_dtypes, exclude_dims=exclude_dims
    758         )
--> 760 result_data = func(*input_data)
    762 if signature.num_outputs == 1:
    763     result_data = (result_data,)

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\xarray\core\duck_array_ops.py:300, in where(condition, x, y)
    298 """Three argument where() with better dtype promotion rules."""
    299 xp = get_array_namespace(condition)
--> 300 return xp.where(condition, *as_shared_dtype([x, y], xp=xp))

File <__array_function__ internals>:200, in where(*args, **kwargs)

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\array\core.py:1761, in Array.__array_function__(self, func, types, args, kwargs)
   1758 if has_keyword(da_func, "like"):
   1759     kwargs["like"] = self
-> 1761 return da_func(*args, **kwargs)

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\array\routines.py:2107, in where(condition, x, y)
   2105     return broadcast_to(out, shape).astype(dtype)
   2106 else:
-> 2107     return elemwise(np.where, condition, x, y)

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\array\core.py:4818, in elemwise(op, out, where, dtype, name, *args, **kwargs)
   4813     need_enforce_dtype = any(
   4814         not is_scalar_for_elemwise(a) and a.ndim == 0 for a in args
   4815     )
   4817 if not name:
-> 4818     name = f"{funcname(op)}-{tokenize(op, dtype, *args, where)}"
   4820 blockwise_kwargs = dict(dtype=dtype, name=name, token=funcname(op).strip("_"))
   4822 if where is not True:

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\base.py:930, in tokenize(*args, **kwargs)
    921 def tokenize(*args, **kwargs):
    922     """Deterministic token
    923 
    924     >>> tokenize([1, 2, '3'])
   (...)
    928     True
    929     """
--> 930     hasher = _md5(str(tuple(map(normalize_token, args))).encode())
    931     if kwargs:
    932         hasher.update(str(normalize_token(kwargs)).encode())

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\utils.py:642, in Dispatch.__call__(self, arg, *args, **kwargs)
    638 """
    639 Call the corresponding method based on type of argument.
    640 """
    641 meth = self.dispatch(type(arg))
--> 642 return meth(arg, *args, **kwargs)

File ~\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\base.py:1219, in register_numpy.<locals>.normalize_array(x)
   1217 else:
   1218     try:
-> 1219         data = hash_buffer_hex(x.ravel(order="K").view("i1"))
   1220     except (BufferError, AttributeError, ValueError):
   1221         data = hash_buffer_hex(x.copy().ravel(order="K").view("i1"))

MemoryError: Unable to allocate 117. GiB for an array with shape (15652569600,) and data type float64
```

Can you please provide insights on how to handle it?

Many Thanks in advance.
Also, how can I apply dask chunks on this.
My data looks like this:
![image](https://user-images.githubusercontent.com/7091088/236663969-02137ce6-97da-4b22-8ce9-1d97553c3d0f.png)

**Comment by jypeter** [comment]
`117. GiB` is indeed quite big, and would be a problem even on a Linux cluster!

I would recommend checking if you can get the results you need with smaller datasets, or rewriting your script and algorithm to loop on smaller subsets of the data

On Linux, you could also try to use [cdo](https://code.mpimet.mpg.de/projects/cdo) to reduce your data before loading it in your python script. They seem to provide some [Win binaries](https://code.mpimet.mpg.de/projects/cdo/files) that you could try to use

**Reply by gkb999** [reply]
> `117. GiB` is indeed quite big, and would be a problem even on a Linux cluster!
> 
> I would recommend checking if you can get the results you need with smaller datasets, or rewriting your script and algorithm to loop on smaller subsets of the data
> 
> On Linux, you could also try to use [cdo](https://code.mpimet.mpg.de/projects/cdo) to reduce your data before loading it in your python script. They seem to provide some [Win binaries](https://code.mpimet.mpg.de/projects/cdo/files) that you could try to use

Thanks for this, but I have clipped the global data to just a small box with 20 latitudes. However, chunking the data as mentioned here helped and produced the temporal climatology.
Thanks for the time again.

**Comment by mfwehner** [comment]
I have this problem a lot.
If it was me, I would divide up the calculation spatially. If latitude-longitude data, I would load one latitude line in at a time via a loop.
Michael

**Reply by pochedls** [reply]
I think this is a good solution that could be implemented pretty easily. 

I was hoping that this could be done relatively seamlessly with dask. I tried this: 

```python
import xcdat as xc                                                                                                                                                
fn = '/p/css03/esgf_publish/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3-Veg/historical/r14i1p1f1/3hr/pr/gr/v20200925/'                                               
ds = xc.open_mfdataset(fn)
# note that the data has 256 latitude and 512 longitude values
# so I am trying to chunk down to very small amounts of data
dsc = ds.chunk(chunks={'lat': 256, 'lon': 512})                                                                                                                   
ds2 = dsc.temporal.climatology("pr", freq="day")   
```

This ended up taking a large amount of memory (so I don't think the calculations are being done at individual grid cells). I'm wondering if @tomvothecoder or @jasonb5 know how to have `xcdat`/`xarray` operate on individual chunks (with a single thread). Do we need to use `map_blocks` with a single thread calculation ([xarray dask documentation](https://docs.xarray.dev/en/stable/user-guide/dask.html))?

**Reply by gkb999** [reply]
Thanks for this, but the climatology is never successful on a daily scale.
When I try to plot the climatology I get the following error.
![image](https://github.com/xCDAT/xcdat/assets/7091088/d2d16b1d-ebbf-4bdf-b2a8-03d2810995a3)


And, the following error is something quite iterative: 
TypeError: float() argument must be a string or a real number, not 'cftime._cftime.DatetimeGregorian'.
While loading the data , it gave following warning using xcdat
![image](https://github.com/xCDAT/xcdat/assets/7091088/e037393d-401b-4fe8-bacb-8a2db12d266b)
How do I resolve this pls?

**Reply by pochedls** [reply]
Hi @GIRIJA-KALYANI – when you post error messages like this, it would be helpful if you could post _text_ (not a screenshot) that boils down the issue to a minimal reproducible example (e.g., [here](https://github.com/xCDAT/xcdat/issues/463)) or at least the lines of code preceding the error message (which is included in your post here). 

If you look at the TypeError message – it seems to have an issue interpreting time. Could you try omitting this, e.g., `plt.plot(rsp.mean(axis=(1,2)))`? If that works, you can create a time vector that matplotlib could interpret. 

Regarding the warnings – you could try to set a calendar attribute (e.g., to `standard`) before doing your operations. I think these are stored in `ds.time.attrs`.

**Reply by gkb999** [reply]
Thanks alot.
`plt.plot(rsp.mean(axis=(1,2)))` : this worked.
Sorry for not posting text and posting a screenshot.
Also, the warning says:
```
2023-05-11 10:17:34,972 [WARNING]: dataset.py(decode_time:341) >> 'time' does not have a calendar attribute set. 
```
Defaulting to CF 'standard' calendar.

So, does it mean it has by default used the standard calendar now?

My ds.time.attrs display:
```
{'long_name': 'Time',
 'delta_t': '0000-00-01 00:00:00',
 'avg_period': '0000-00-01 00:00:00',
 'axis': 'T',
 'actual_range': array([66352., 66473.]),
 'bounds': 'time_bnds'}
```
(this is for the climatology generated from the temporal climatology function of xcdat)

**Reply by pochedls** [reply]
NetCDF writes time in units like "days since 2000-01-01" (the unit, days, and reference time, 2000-01-01, can change). The interpretation of time depends on the [calendar](https://unidata.github.io/cftime/api.html) embedded in the dataset (e.g., proleptic_gregorian, 360_day, etc.). For this reason, it is important to know the calendar. Your dataset must not include the calendar. This warning message basically says that xcdat is assuming that you are using a `standard` calendar. To get rid of the warning, you need to specify the calendar. You can do this with something like (just make sure `ds.time.attrs` includes the calendar when you're done):

```python
time = ds.time
attrs = time.attrs
attrs['calendar'] = 'standard'
ds['time'] = time
```

**Reply by gkb999** [reply]
Thanks, my original data looks like: (attaching screehshot just because it can give a glimpse of format)
![image](https://github.com/xCDAT/xcdat/assets/7091088/d2392ccf-0667-4536-99be-0e5be715060b)
Even after defining the attrs to : attrs['calendar'] = 'standard', it looks similar
![image](https://github.com/xCDAT/xcdat/assets/7091088/1f22863a-0039-4cea-83a1-40c559d97b93)

Also in the first screenshot, although the warning appeared, xcdat_boundas are True(ref: screenshot 1). After assigning, I cannot see that xcdat_bounds thing in my data variables list.

Second thing:
I want to calculate the daily and seasonal anomalies, based on the temporal climatology I derived using xcdat.
For seasonal anaomalies:
I tried: 
```python
dsc_seas = dsc.temporal.climatology(
    "icec",
    freq="season",
    weighted=True,
    season_config={"dec_mode": "DJF", "drop_incomplete_djf": True},
)
```
and my dsc_seas.icec.time array looks like:
```
array.DataArray'time'time: 4
array([cftime.DatetimeGregorian(1, 1, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1, 4, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1, 7, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1, 10, 1, 0, 0, 0, 0, has_year_zero=False)],
      dtype=object)
Coordinates:
time
(time)
object
0001-01-01 00:00:00 ... 0001-10-...
Indexes: (1) 
```
Ref: 
![image](https://github.com/xCDAT/xcdat/assets/7091088/f51fe8e4-b7fa-4c76-8fd8-6df11992ba9d)

Calendar is standard here.
However, when I plot: it shows:
```
RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args)) and return empty figures with no data.
```
What could be the error?

**Reply by gkb999** [reply]
Okay, it could be becuase the code seems to divide something by zero/nan.
It is highly possible, as I'm dealing with sea-ice areas here and most of them around summer would be 0.
I'll try `np.seterr(divide='ignore', invalid='ignore')`.

**Reply by pochedls** [reply]
@GIRIJA-KALYANI - I don't fully understand the first probelm – you were setting the calendar to get rid of the warning message. Presumably the calendar you set shows up in `ds.time.attrs` now. I think you may have figured out the second issue.

This thread is becoming a little difficult to follow. I would really appreciate if you shared [formatted, reproducible code snippets](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) and briefly summarize what the problem is and what you tried to do to resolve the problem. If you searched for similar errors and tried things from stackoverflow or looked at matplotlib/xarray documentation, you can link to those (and share code that did not help).

**Comment by gkb999** [comment]
I'm trying to replicate the seasonal climatologies as [https://github.com/xCDAT/xcdat/blob/main/docs/examples/climatology-and-departures.ipynb
](url)
My dataset:
```
xarray.Dataset
Dimensions:
time: 15097lat: 28lon: 160bnds: 2
Coordinates:
time
(time)
object
1981-09-01 00:00:00 ... 2022-12-...
lat
(lat)
float32
-79.88 -79.62 ... -73.38 -73.12
lon
(lon)
float32
160.1 160.4 160.6 ... 199.6 199.9
Data variables:
icec
(time, lat, lon)
float32
dask.array<chunksize=(122, 28, 160), meta=np.ndarray>
lon_bnds
(lon, bnds)
float32
dask.array<chunksize=(160, 2), meta=np.ndarray>
lat_bnds
(lat, bnds)
float32
dask.array<chunksize=(28, 2), meta=np.ndarray>
time_bnds
(time, bnds)
object
dask.array<chunksize=(15097, 2), meta=np.ndarray>
```

For seasonal climatology: I tried the follwoing 
```python
dsc_seas = dsc.temporal.climatology(
    "icec",
    freq="season",
    weighted=True,
    season_config={"dec_mode": "DJF", "drop_incomplete_djf": True},
)
```

which returns the follwoing error:
```
C:\Users\buradagi\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\core.py:119: RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args))
```

My seasonal climatology data (given below) although returned 4 time variables (4 seasons), while plotting displays the Runtime Warning with blank plots
```
xarray.Dataset
Dimensions: lat: 28lon: 160bnds: 2time: 4

Coordinates: 
lat
(lat)
float32
-79.88 -79.62 ... -73.38 -73.12
lon
(lon)
float32
160.1 160.4 160.6 ... 199.6 199.9
time
(time)
object
0001-01-01 00:00:00 ... 0001-10-...

Data variables:
lon_bnds
(lon, bnds)
float32
dask.array<chunksize=(160, 2), meta=np.ndarray>
lat_bnds
(lat, bnds)
float32
dask.array<chunksize=(28, 2), meta=np.ndarray>
icec
(time, lat, lon)
float64
dask.array<chunksize=(1, 28, 160), meta=np.ndarray>

long_name : Daily Sea Ice Concentration
units : percent
valid_range : [0. 1.]
precision : 2.0
dataset : NOAA High-resolution Blended Analysis
var_desc : Sea Ice Concentration
level_desc : Surface
statistic : Mean
parent_stat : Individual Observations
actual_range : [0.01 1.  ]
operation : temporal_avg
mode : climatology
freq : season
weighted : True
dec_mode : DJF
drop_incomplete_djf : True
```


**My Goal: Seasonal nd Daily anomalies using the xcdat temporal climatology.
Thanks in advance**

**Reply by pochedls** [reply]
@GIRIJA-KALYANI - could you try to figure out if this is a plotting issue or a data issue? I appreciate the additional context, but it would be extremely helpful if you could provide a _complete minimal example_ leading to the issue (see the example below or a [minimal complete verifiable example](https://github.com/xCDAT/xcdat/issues/463). From your notes above, I can't really tell if there was a problem in calculating the seasonal climatology or a problem in plotting something (you don't share your plot command or what `xarray.Dataset` you are printing out). 

### example file
```bash
wget https://downloads.psl.noaa.gov/Datasets/noaa.oisst.v2.highres/icec.day.mean.1982.nc
```

### python example
```python
# imports
import xcdat as xc
import matplotlib.pyplot as plt
# open dataset
ds = xc.open_dataset('icec.day.mean.1982.nc')
# climatology calculation
dsc = ds.temporal.climatology("icec", freq="season", weighted=True, season_config={"dec_mode": "DJF", "drop_incomplete_djf": True})
# plot
plt.pcolor(dsc.icec[0])
plt.colorbar()
```
![Screenshot 2023-05-10 at 9 40 16 PM](https://github.com/xCDAT/xcdat/assets/3698109/3ea40866-8bb4-4223-8e64-6c057eae60fb)

**Reply by gkb999** [reply]
Thanks for this, I'm using the same data. For 40 years!!! 
```python
ice_daily = xc.open_mfdataset('/icec/*.nc', combine = 'nested', concat_dim="time")  # reads all the files in the folder
```

Although gives the following error:
```
023-05-11 10:17:34,687 [WARNING]: dataset.py(decode_time:341) >> 'time' does not have a calendar attribute set. Defaulting to CF 'standard' calendar.
2023-05-11 10:17:34,694 [WARNING]: dataset.py(decode_time:341) >> 'time' does not have a calendar attribute set. Defaulting to CF 'standard' calendar.
2023-05-11 10:17:34,700 [WARNING]: dataset.py(decode_time:341) >> 'time' does not have a calendar attribute set. Defaulting to CF 'standard' calendar.
```

Does the job perfectly as we can see time:15097
```
<xarray.Dataset>
Dimensions:    (time: 15097, lat: 720, lon: 1440, bnds: 2)
Coordinates:
  * time       (time) object 1981-09-01 00:00:00 ... 2022-12-31 00:00:00
  * lat        (lat) float32 -89.88 -89.62 -89.38 -89.12 ... 89.38 89.62 89.88
  * lon        (lon) float32 0.125 0.375 0.625 0.875 ... 359.1 359.4 359.6 359.9
Dimensions without coordinates: bnds
Data variables:
    icec       (time, lat, lon) float32 dask.array<chunksize=(122, 720, 1440), meta=np.ndarray>
    lon_bnds   (lon, bnds) float32 0.0 0.25 0.25 0.5 ... 359.5 359.8 359.8 360.0
    lat_bnds   (lat, bnds) float32 -90.0 -89.75 -89.75 ... 89.75 89.75 90.0
    time_bnds  (time, bnds) object 1981-08-31 12:00:00 ... 2022-12-31 12:00:00
```

Now, clipped to my ROI:
```python
rs = ds.sel(lon = slice(160, 200),lat =slice(-80, -73))
```

Looks like:
```
<xarray.Dataset>
Dimensions:    (time: 15097, lat: 28, lon: 160, bnds: 2)
Coordinates:
  * time       (time) object 1981-09-01 00:00:00 ... 2022-12-31 00:00:00
  * lat        (lat) float32 -79.88 -79.62 -79.38 ... -73.62 -73.38 -73.12
  * lon        (lon) float32 160.1 160.4 160.6 160.9 ... 199.1 199.4 199.6 199.9
Dimensions without coordinates: bnds
Data variables:
    icec       (time, lat, lon) float32 dask.array<chunksize=(122, 28, 160), meta=np.ndarray>
    lon_bnds   (lon, bnds) float32 160.0 160.2 160.2 160.5 ... 199.8 199.8 200.0
    lat_bnds   (lat, bnds) float32 -78.0 -78.0 -78.0 ... -73.25 -73.25 -73.0
    time_bnds  (time, bnds) object 1981-08-31 12:00:00 ... 2022-12-31 12:00:00
Attributes:
    Conventions:    CF-1.5
    title:          NOAA High-resolution Blended Analysis: Daily Values using...
```

Temporal Climatology day wise:
```python
dsclim = dsc.temporal.climatology("icec", freq="day") #worked fine
```

but when I want to display:
It gives the follwoing error:
```
"\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\core.py:119: RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args))"
with no output
<matplotlib.collections.PolyCollection at 0x20bfcf5cd10>
```

The same for seasonal climatology as well:
Code is below:
```python
dsc_seas = dsc.temporal.climatology(
    "icec",
    freq="season",
    weighted=True,
    season_config={"dec_mode": "DJF", "drop_incomplete_djf": True},
)
```
but cant  visualise it.
returns the same Runtime error again:
```
\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\core.py:119: RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args))
\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\core.py:119: RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args))
\Anaconda3\envs\xcdat_env\Lib\site-packages\dask\core.py:119: RuntimeWarning: invalid value encountered in divide
  return func(*(_execute_task(a, cache) for a in args))
```
All I want is: solve the plotting issue.
Many thanks.

I tried putting up the issue the best way I can.
Is it because I'm using files mfdataset to concatenate ?

This issue was not before though.

**Reply by gkb999** [reply]
Well, after restarting the system, although the error persists, I can view the maps.
Thanks a lot for the assistance.

**Reply by pochedls** [reply]
That is great news – thank you for following up. I was confused, because these are _warnings_ (not _errors_) so it seems like the calculation is (or should be) completing. I think there may be lingering memory issues since this dataset is so large – this may be why the restart helped. I think we're going to explore xcdat + dask in the future, so we may add more documentation about this. We'll keep in mind large datasets as something to contend with.

**Reply by lee1043** [reply]
Thank you for the discussion. I added some minor edits to some of text from @GIRIJA-KALYANI, formatting code and log messages for easier read of this thread for the future reference.

**Comment by gkb999** [comment]
That would be great, seriously.
Climate data generally deals with such huge chunks, but xcdat provides
easier and accurate solutions.
Exploring other API too from the module.

Thanks alot for the response again.




*Girija Kalyani BuradaDoctoral StudentSchool of Geography, Environment and
Earth SciencesVictoria University of WellingtonNew Zealand*

**Reply by tomvothecoder** [reply]
Awesome, glad to hear you got things working and `xcdat` is helping you with your work (especially with large datasets).
Stay tuned for a Dask guide that will be in the available through the docs in the near future.
